server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java

```diff
@@ -383,27 +393,71 @@ protected Query doToQuery(QueryShardContext context) {
+                // before, including creating lucene fieldcache (!)
+                // in this case, execute disjoint as exists && !intersects
+                BooleanQuery.Builder bool = new BooleanQuery.Builder();
+                Query exists = ExistsQueryBuilder.newFilter(context, fieldName);
+                Query intersects = prefixTreeStrategy.makeQuery(getArgs(shapeToQuery, ShapeRelation.INTERSECTS));
+                bool.add(exists, BooleanClause.Occur.MUST);
+                bool.add(intersects, BooleanClause.Occur.MUST_NOT);
+                query = new ConstantScoreQuery(bool.build());
+            } else {
+                query = new ConstantScoreQuery(prefixTreeStrategy.makeQuery(getArgs(shapeToQuery, relation)));
+            }
         } else {
-            query = new ConstantScoreQuery(strategy.makeQuery(getArgs(shapeToQuery, relation)));
+            query = getVectorQuery(context, shapeFieldType, shapeToQuery);
         }
         return query;
     }
 
+    private Query getVectorQuery(QueryShardContext context, GeoShapeFieldMapper.GeoShapeFieldType shapeFieldType,
+                                 ShapeBuilder queryShapeBuilder) {
+        // CONTAINS queries are not yet supported by VECTOR strategy
+        if (relation == ShapeRelation.CONTAINS) {
+            throw new QueryShardException(context,
+                ShapeRelation.CONTAINS + " query relation not supported for Field [" + fieldName
+                    + "] when using Strategy [" + shapeFieldType.strategy() + "]");
+        }
+
+        Object queryShape = queryShapeBuilder.buildLucene();
+        Query geoQuery;
+        if (queryShape instanceof Line[]) {
+            geoQuery = LatLonShape.newLineQuery(fieldName(), relation.getLuceneRelation(), (Line[]) queryShape);
+        } else if (queryShape instanceof Polygon[]) {
+            geoQuery = LatLonShape.newPolygonQuery(fieldName(), relation.getLuceneRelation(), (Polygon[]) queryShape);
+        } else if (queryShape instanceof Line) {
+            geoQuery = LatLonShape.newLineQuery(fieldName(), relation.getLuceneRelation(), (Line) queryShape);
+        } else if (queryShape instanceof Polygon) {
+            geoQuery = LatLonShape.newPolygonQuery(fieldName(), relation.getLuceneRelation(), (Polygon) queryShape);
+        } else if (queryShape instanceof Rectangle) {
+            Rectangle r = (Rectangle) queryShape;
+            geoQuery = LatLonShape.newBoxQuery(fieldName(), relation.getLuceneRelation(),
+                r.minLat, r.maxLat, r.minLon, r.maxLon);
+        } else {
+            // geometry types not yet supported by Lucene's LatLonShape
+            GeoShapeType geometryType = null;
+            if (queryShape instanceof double[][]) {
+                geometryType = GeoShapeType.MULTIPOINT;
+            } else if (queryShape instanceof double[] || queryShape instanceof GeoPoint) {
+                geometryType = GeoShapeType.POINT;
```

Comment:
I think we can offer a point query by reusing the Box query. We just need to make minLat == maxLat and minLon == maxLon.

Reply:
:+1:  I'll run some tests with that approach.  I was also considering implementing `GeometryCollection` query as a `BooleanQuery`; though we'll likely need to run some benchmarks to make sure performance is acceptable.

Reply:
Implemented this for `POINT` queries, and it will work quite nicely. Unfortunately I found (and fixed) a small bug in `LatLonShapeBoundingBoxQuery`. So I'll either need to upgrade the lucene version to commit the implementation or hold off until another PR.
I do not want to implement this approach for `MULTIPOINT` queries, however. It would require rewriting to a `BooleanQuery` which would be quite slow, and abusive, for even just a few points.